[Vikrant Prakash] iP

viki0526 opened the pull request:
### Johnny

> "Manual To-do lists are very tiresome" - Me (and probably you too)

Johnny collects and organises your tasks, deadlines and tasks for you so that you can focus on the more important things!

It is... 
- Command-based
- **Super Easy** to learn (Refer to [User Guide](https://nus-cs2103-ay2122s2.github.io/website/schedule/week4/project.html))
- FREE 💯

Features:
1. Managing tasks
2. Managing deadlines and events
3. Reminders

Improvements Planned
- [x] Modifying tasks
- [ ] GUI (Coming Soon)

If you are a Java programmer, you can use it to practice Java too. Here's the main method:
```
public static void main(String[] args) throws InvalidArgumentsException,
        EmptyDescriptionException, NoDateException  {
        new Johnny("tasklist.txt").run();
}
```

Commit:
Add Gradle support

viki0526 committed:
Added Greet, Echo and Exit Functionality

viki0526 committed:
Removed Duke.java

viki0526 committed:
Added list functionality

viki0526 committed:
Added functionality to mark tasks done or not done yet

viki0526 committed:
Added Todo, Deadline and Event classification

viki0526 committed:
Added Automated TextUI Testing

viki0526 committed:
Added Exception Handling for Invalid Arguments and Empty Description/No Date

viki0526 committed:
Added delete command

viki0526 committed:
Fix error handling

viki0526 committed:
Add Text File handling of tasklist

viki0526 committed:
Add Date handling using Java.LocalDate

viki0526 committed:
Merge branch 'master' into branch-Level-7

* master:
  Add Date handling using Java.LocalDate

viki0526 committed:
Add OOP for input parsing and text file operations

viki0526 committed:
Put in classes into package 'johnny'

viki0526 committed:
Add JUnit tests for Parser.parse() and Task.toString()

viki0526 committed:
Add JavaDoc

viki0526 committed:
Add Coding Standard improvements

viki0526 committed:
Add Find Feature to search for tasks

viki0526 committed:
Merge branch 'branch-A-CodingStandard'

* branch-A-CodingStandard:
  Add Coding Standard improvements

# Conflicts:
#	src/main/java/johnny/InputList.java

viki0526 committed:
Merge branch 'branch-Level-9'

* branch-Level-9:
  Add Find Feature to search for tasks

# Conflicts:
#	src/main/java/johnny/InputList.java

viki0526 committed:
Level-9 Part 2

Kidsnd274 commented on src/main/java/johnny/Parser.java:
```diff
@@ -0,0 +1,111 @@
+        String[] tags = input.split(" ", 2);
+        ArrayList<String> result = new ArrayList<>();
+
+        if(input.equals("bye")) {
```
A switch-case statement here may make the code more neater

Kidsnd274 commented on src/main/java/johnny/InvalidArgumentsException.java:
```diff
@@ -0,0 +1,21 @@
+package johnny;
+
+public class InvalidArgumentsException extends Exception{
```
Nice usage of exceptions! I probably should use more exceptions in my project.

Korochii replied:
```suggestion
public class InvalidArgumentsException extends Exception {
```
Missing whitespace between Exception and the curly brace.

Kidsnd274 commented on src/test/java/ParserTest.java:
```diff
@@ -0,0 +1,20 @@
+
+import java.util.ArrayList;
+
+public class ParserTest {
```
Nice use of test cases! I probably should add more of them in my project.

Kidsnd274 commented on src/main/java/johnny/Parser.java:
```diff
@@ -0,0 +1,111 @@
+        try {
+            Integer.parseInt(str);
+            return true;
+        } catch(NumberFormatException e){
```
There should be a spacing between catch and "NumberFormatException"

Korochii replied:
```suggestion
        } catch (NumberFormatException e) {
```
I agree with Kidsnd274 too! In addition, there should be a whitespace between ( and {

Kidsnd274 reviewed (commented):
Nice job on this project! Didn't really have much to comment on so I was scraping the barrel on comments. Just some nitpick comments.

Korochii commented on src/main/java/johnny/EmptyDescriptionException.java:
```diff
@@ -0,0 +1,14 @@
+package johnny;
+
+public class EmptyDescriptionException extends Exception{
```
```suggestion
public class EmptyDescriptionException extends Exception {
```
Missing whitespace between Exception and the curly brace.

Korochii commented on src/main/java/johnny/Deadline.java:
```diff
@@ -0,0 +1,37 @@
+
+    @Override
+    public String toString() {
+        return "[D] [" + this.getStatusIcon() + "] " +
+                super.getDescription() + " (" +
+                this.getDueDate().format(DateTimeFormatter.ofPattern("MMM dd yyyy")) + ")";
```
```suggestion
        return "[D] [" + this.getStatusIcon() + "] " 
                + super.getDescription() + " (" 
                + this.getDueDate().format(DateTimeFormatter.ofPattern("MMM dd yyyy")) + ")";
```
I like how the string is represented here as it is easy to understand the different parts/elements that are in the string.
However, there is a slight area of improvement, which is to place the line break before the "+" operator.

Korochii commented on src/main/java/johnny/Event.java:
```diff
@@ -0,0 +1,37 @@
+
+    @Override
+    public String toString() {
+        return "[E] [" + this.getStatusIcon() + "] " +
+                super.getDescription() + " (" +
+                this.getEventDate().format(DateTimeFormatter.ofPattern("MMM dd yyyy")) + ")";
```
Same issue as Deadline.java

Korochii commented on src/main/java/johnny/NoDateException.java:
```diff
@@ -0,0 +1,15 @@
+package johnny;
+
+public class NoDateException extends Exception{
```
```suggestion
public class NoDateException extends Exception {
```
Missing whitespace between Exception and the curly brace.

Korochii commented on src/main/java/johnny/Parser.java:
```diff
@@ -0,0 +1,111 @@
+     */
+    public ArrayList<String> parse() throws EmptyDescriptionException, NoDateException, InvalidArgumentsException {
+        String[] tags = input.split(" ", 2);
+        ArrayList<String> result = new ArrayList<>();
```
```suggestion
        ArrayList<String> results = new ArrayList<>();
```
Great job in using an ArrayList to store the results here! However, I think plural form should be used on names representing a collection of objects.

Korochii commented on src/main/java/johnny/Storage.java:
```diff
@@ -0,0 +1,73 @@
+    public ArrayList<Task> readTasks() throws IOException{
+        ArrayList<Task> taskList = new ArrayList<>();
+
+            File file = new File(
+                    filepath);
```
Might want to condense these lines into a single line for easier readability.

Korochii commented on src/main/java/johnny/Parser.java:
```diff
@@ -0,0 +1,111 @@
+                throw new EmptyDescriptionException();
+            }
+            result.add(ADD_TODO);
+            result.add(content);
+        }
+        else if(tags[0].equals("deadline")) {
+            if(tags.length == 1 || tags[1].equals("")) {
+                throw new EmptyDescriptionException();
+            }
+
+            if(!tags[1].contains("/")) {
+                throw new NoDateException();
+            }
+
+            String content = tags[1];
+            String[] details = content.split("/", 2);
+
+            result.add(ADD_DEADLINE);
+            result.add(details[0]);
+            result.add(details[1]);
+        }
+        else if(tags[0].equals("event")) {
+            if(tags.length == 1 || tags[1].equals("")) {
+                throw new EmptyDescriptionException();
+            }
+
+            if(!tags[1].contains("/")) {
+                throw new NoDateException();
+            }
+
+            String content = tags[1];
+            String[] details = content.split("/", 2);
+
+            result.add(ADD_EVENT);
+            result.add(details[0]);
+            result.add(details[1]);
+        }
+        else {
+            throw new InvalidArgumentsException(input);
+        }
```
I like how you separate each cases clearly! However, there are missing white spaces between each "else if" and the next bracket.
In addition, some of the else/else if are not on the same line as the previous closing curly brace (such as line 89-90, 105-106).

Korochii commented on src/main/java/johnny/Ui.java:
```diff
@@ -0,0 +1,88 @@
+            else if(commandTag.equals(Parser.FIND_EVENT)) {
+                System.out.println("Here are the matching tasks in your list:");
+                tasks.searchEventAndPrint(parseOutput.get(1));
+            }
+            else if(commandTag.equals(Parser.ADD_TODO)) {
+
+                Task newTask = new Todo(parseOutput.get(1), false);
+                tasks.add(newTask);
+
+                System.out.println("Got it! I've added this task:");
+                System.out.println(newTask);
+                System.out.println("Now you have " + tasks.getCount() + " tasks in your list.");
+            }
+            else if(commandTag.equals(Parser.ADD_DEADLINE)) {
+
+                Task newTask = new Deadline(parseOutput.get(1),
+                        LocalDate.parse(parseOutput.get(2)), false);
+
+                tasks.add(newTask);
+
+                System.out.println("Got it! I've added this task:");
+                System.out.println(newTask);
+                System.out.println("Now you have " + tasks.getCount() + " tasks in your list.");
+            }
+            else if(commandTag.equals(Parser.ADD_EVENT)) {
+
+                Task newTask = new Event(parseOutput.get(1),
+                        LocalDate.parse(parseOutput.get(2)), false);
+
+                tasks.add(newTask);
+
+                System.out.println("Got it! I've added this task:");
+                System.out.println(newTask);
+                System.out.println("Now you have " + tasks.getCount() + " tasks in your list.");
+            }
+            else {
+                throw new InvalidArgumentsException(input);
+            }
+        }
+    }
```
Similar indentation/whitespace issue to Parser.java

Korochii commented on src/test/java/ParserTest.java:
```diff
@@ -0,0 +1,20 @@
+
+    @Test
+    public void parse_addTodoTask_success() throws Exception{
+        ArrayList<String> expected = new ArrayList<>();
```
expected should be in plural form

Korochii commented on src/test/java/ParserTest.java:
```diff
@@ -0,0 +1,20 @@
+public class ParserTest {
+
+    @Test
+    public void parse_addTodoTask_success() throws Exception{
```
```suggestion
    public void parse_addTodoTask_success() throws Exception {
```
Missing whitespace between Exception and {

Korochii reviewed (commented):
Overall, there are some minor issues (mostly related to whitespaces/indentations) that should be easy to fix. 
Good job and keep up the good work! :smiley:

viki0526 committed:
Merge branch 'add-gradle-support' of https://github.com/viki0526/ip

* 'add-gradle-support' of https://github.com/viki0526/ip:
  Add Gradle support

viki0526 committed:
Add GUI

viki0526 committed:
Add assertions to various points in the code

There are several methods that could return null values if the code
execution goes wrong. This is probably due to unhandled runtime
exceptions

Adding assert statements allows for easy debugging of Null Pointer
Exceptions

Let's make code debugging easier for future developers using these
assertions

viki0526 committed:
Improve code quality

There are several method and class names that did
not follow naming conventions.

Naming them concisely and descriptively makes the code
much more readable

A lot of dead code has been removed and long methods have
been shortened

As a step towards more understandable, better quality code
let's refactor this code and improve its quality

viki0526 committed:
Merge pull request #3 from viki0526/branch-A-Assertions

Add assertions to various points in the code

viki0526 committed:
Merge pull request #2 from viki0526/branch-A-CodeQuality

Improve code quality

viki0526 committed:
Change profile pictures, delete inessential classes and files

viki0526 committed:
Merge branch 'master' of https://github.com/viki0526/ip

* 'master' of https://github.com/viki0526/ip:
  Improve code quality
  Add assertions to various points in the code

# Conflicts:
#	src/main/java/johnny/Johnny.java

viki0526 committed:
Delete text-ui testing and test files

viki0526 committed:
Add feature to detect scheduling clashes

viki0526 committed:
Add docs for github website

viki0526 committed:
Add docs update

viki0526 committed:
Add docs update #2

viki0526 committed:
Add gradle support
